src/vm/eventtrace.cpp

```diff
@@ -4441,12 +4441,40 @@ extern "C"
 
         BOOLEAN bIsRundownTraceHandle = (context->RegistrationHandle==Microsoft_Windows_DotNETRuntimeRundownHandle);
 
-		// TypeSystemLog needs a notification when certain keywords are modified, so
-		// give it a hook here.
-		if (g_fEEStarted && !g_fEEShutDown && bIsPublicTraceHandle)
-		{
-			ETW::TypeSystemLog::OnKeywordsChanged();
-		}
+        if (g_fEEStarted && !g_fEEShutDown)
+        {
+            // TypeSystemLog needs a notification when certain keywords are modified, so
+            // give it a hook here.
+            if (bIsPublicTraceHandle)
+            {
+                ETW::TypeSystemLog::OnKeywordsChanged();
+            }
+
+            if (ControlCode == EVENT_CONTROL_CODE_ENABLE_PROVIDER || ControlCode == EVENT_CONTROL_CODE_DISABLE_PROVIDER)
+            {
+                static_assert(GCEventLevel_None == TRACE_LEVEL_NONE, "GCEventLevel_None value mismatch");
+                static_assert(GCEventLevel_Fatal == TRACE_LEVEL_FATAL, "GCEventLevel_Fatal value mismatch");
+                static_assert(GCEventLevel_Error == TRACE_LEVEL_ERROR, "GCEventLevel_Error value mismatch");
+                static_assert(GCEventLevel_Warning == TRACE_LEVEL_WARNING, "GCEventLevel_Warning mismatch");
+                static_assert(GCEventLevel_Information == TRACE_LEVEL_INFORMATION, "GCEventLevel_Information mismatch");
+                static_assert(GCEventLevel_Verbose == TRACE_LEVEL_VERBOSE, "GCEventLevel_Verbose mismatch");
+
+                // The GC also needs to be informed of changes to keywords and levels.
+                IGCHeap *heap = GCHeapUtilities::GetGCHeap();
+                GCEventKeyword keywords = static_cast<GCEventKeyword>(MatchAnyKeyword);
```

Comment:
The current implementation supports MatchAllKeywords as well, so you should probably support this as well.  From MSDN:

```
This bitmask is optional. This mask further restricts the category of events that you want the provider to write. If the event's keyword meets the MatchAnyKeyword condition, the provider will write the event only if all of the bits in this mask exist in the event's keyword. This mask is not used if MatchAnyKeyword is zero. See Remarks.
```

Reply:
Do we have any events with multiple keywords? (I'm not aware of any for the GC - not sure if there are others elsewhere.) If I'm reading the documentation correctly, it seems to me that `MatchAllKeywords` is only useful in that case.

Reply:
Here's an example:

```
                   <event value="11" version="1" level="win:Informational"  template="GCCreateConcurrentThread"
                           keywords ="GCKeyword ThreadingKeyword"  opcode="GCCreateConcurrentThread"
                           task="GarbageCollection"
                           symbol="GCCreateConcurrentThread_V1" message="$(string.RuntimePublisher.GCCreateConcurrentThread_V1EventMessage)"/>
```

I suspect it's not a huge deal.  LTTng and EventPipe don't have this concept, but I don't want us to inadvertently miss it.

@vancem, do you know how big of a deal it is to not support MatchAllKeywords for Local GC?

Reply:
If we do want to support MatchAllKeywords I think we'll need some additional design work. The GC would need to keep track of both the Any and All keyword masks in order to stay correct.